vic/drivers/shared_all/src/calc_root_fraction.c

```diff
@@ -41,126 +41,121 @@ calc_root_fractions(veg_con_struct  *veg_con,
-        Lsum = Lstep;
-        Zsum = 0;
         zone = 0;
+        Lsum = soil_con->depth[layer];
+        Lsumprev = 0;
+        Zsum = veg_con[veg].zone_depth[zone];
+        Dsum = 0;
+        Dsumprev = 0;
+        sum_depth = 0;
+        sum_dens = 0;
 
-        n_iter = 0;
-        while (zone < options.ROOT_ZONES) {
-            n_iter++;
-            if (n_iter > MAX_ROOT_ITER) {
-                log_warn("veg=%d of Nveg=%d", veg, Nveg);
-                log_warn("zone %zu of %zu ROOT_ZONES", zone,
-                         options.ROOT_ZONES);
-                log_err("stuck in an infinite loop");
-            }
+        while (layer < options.Nlayer || zone < options.ROOT_ZONES) {
 
-            Zstep = veg_con[veg].zone_depth[zone];
-            if ((Zsum + Zstep) <= Lsum && Zsum >= Lsum - Lstep) {
-                /** CASE 1: Root Zone Completely in Soil Layer **/
-                sum_fract += veg_con[veg].zone_fract[zone];
+            // Compute depth interval for consideration
+            if (Lsum <= Zsum && layer < options.Nlayer) {
+                Dsum = Lsum;
             }
             else {
-                /** CASE 2: Root Zone Partially in Soil Layer **/
-                if (Zsum < Lsum - Lstep) {
-                    /** Root zone starts in previous soil layer **/
-                    Zmin_depth = Lsum - Lstep;
-                    Zmin_fract = linear_interp(Zmin_depth, Zsum, Zsum + Zstep,
-                                               0,
-                                               veg_con[veg].zone_fract[zone]);
-                }
-                else {
-                    /** Root zone starts in current soil layer **/
-                    Zmin_depth = Zsum;
-                    Zmin_fract = 0.;
+                Dsum = Zsum;
+            }
+            dD = Dsum - Dsumprev;
+            Dsumprev = Dsum;
+
+            // Add to running totals
+            sum_depth += dD;
+            if (zone < options.ROOT_ZONES) {
+                sum_dens += dD * root_dens[zone];
+            }
+
+            dL = Lsum - Lsumprev;
+
+            // Compute weighted integral of root densities over soil layer
+            // Wait to do this until either we've completed a soil layer
+            // or we're at the final layer and we've completed all root zones
+            if ( dL > 0 && Dsum == Lsum &&
+                 ( layer < options.Nlayer - 1 ||
+                   ( layer >= options.Nlayer - 1 &&
+                     zone >= options.ROOT_ZONES - 1 &&
+                     Dsum == Zsum ) ) ) {
+                ltmp = layer;
+                if (layer >= options.Nlayer - 1) {
+                    ltmp = options.Nlayer - 1;
                 }
-                if (Zsum + Zstep <= Lsum) {
-                    /** Root zone ends in current layer **/
-                    Zmax = Zsum + Zstep;
+                if (sum_depth > 0) {
+                    veg_con[veg].root[ltmp] = dL * sum_dens / sum_depth;
```

Comment:
It seems like `sum_dens` already accumulates the fraction in one sol layer, so why do we need to apply further rescaling here? (After chatting with Ted offline we think `sum_dens` and `sum_depth` may be the same in most cases, and the code could be simplified)

Reply:
Correction: we think `sum_depth` and `dL` may be the same - and might therefore always cancel out.  I originally set up the code this way because I thought I'd need to allow `sum_depth` and `dL` to be unequal, e.g., if the bottom of the soil column and the bottom of the root zones don't match, but the way the code ended up, I think `Lsum` and `Zsum` end up being set equal to each other in these cases, meaning that `sum_depth` and `dL` would end up being equal all the time.  I'm going to test out a simplification of the code and upload it if it works.  Stay tuned...